Define the ACHL cache and ACHL/ACH processing

yoavweiss opened the pull request:
This PR replaces https://github.com/whatwg/fetch/pull/729 in resolving https://github.com/whatwg/fetch/issues/726, which as [discussed](https://github.com/whatwg/fetch/issues/726#issuecomment-391676446) should move to HTML. Rough design sketch is also [available](https://docs.google.com/document/d/1eBMLVHO3p1hwltL9xJJ5LwViW1-yqOvS0w6SNioOUXc/edit#).

This initial PR still doesn't resolve everything that was previously discussed (e.g. doesn't clear the ACHL cache along with cookies just yet), but I'd love to get an early opinion on it, to see if this is the right direction.

domenic commented on source:
```diff
@@ -82317,6 +82412,26 @@ interface <dfn>Location</dfn> { // but see also <a href="#the-location-interface
 
+     <li><p>Let <var>callerIsSecureContext</var> be the result of executing <span>Is environment
+     settings object a secure context?</span> on the <code>Window</code> object's
+     environment settings object.</p></li>
```
This should be the Document's relevant settings object. "The Window" is not clear; a Window object is only created in some cases during step 2.

You may want to save this since in a variable since you use it in the below steps.

yoavweiss replied:
ack

domenic commented on source:
```diff
@@ -82317,6 +82412,26 @@ interface <dfn>Location</dfn> { // but see also <a href="#the-location-interface
+     Accept-CH header</a> is present, and <var>callerIsSecureContext</var> is true, add each parsed
+     client hints token to the <span data-x="concept-settings-object-client-hints-list">
+     client-hints list</span> of the <code>Window</code> object's environment settings object.</p>
+     </li>
```
How does the parsing work?

Also, typographically, use the style above from Referrer-Policy (backticks and code)

yoavweiss replied:
The ABNF for parsing is defined at http://httpwg.org/http-extensions/client-hints.html#accept-ch
I guess I could refer to "#field-name" here, or if needed, define an explicit parsing algorithm.

domenic replied:
I wasn't able to to find a definition of `#field-name`, but if you think implementers will be able to figure this out unambiguously (especially with regard to different types of whitespace, etc.) based on their knowledge of HTTP specs, maybe it'll be good enough.

domenic replied:
Although, assuming `#field-name` is defined via an ABNF, that only gives you a yes/no answer, not a list of strings (or is it a list of byte sequences). So maybe deferring to https://infra.spec.whatwg.org/#split-on-commas would be a good idea? Probably after doing either a UTF-8 or an isomorphic decode?

domenic commented on source:
```diff
@@ -86177,8 +86292,59 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
+  <li><dfn data-x="concept-accept-ch-lifetime-cache-max-age"
+  data-definition-for="concept-accept-ch-lifetime-cache">max-age</dfn> (a number of seconds)</li>
+  <li><dfn data-x="concept-accept-ch-lifetime-cache-client-hints-list"
+  data-definition-for="concept-accept-ch-lifetime-cache">client-hints-list</dfn></li>
```
"(a list of X)" would be good

yoavweiss replied:
I thought it should match (or eventually replace) https://fetch.spec.whatwg.org/#client-hints-list

domenic replied:
Oh interesting. Then it should probably be "(a client hints list)" which links to that definition. Although, that definition might be best changed to a set, not a list.

domenic commented on source:
```diff
@@ -86177,8 +86292,59 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
+  storing the entry. Entries may be removed before that.
+
+  <p>To <dfn data-x="accept-ch-lifetime-cache-create-entry"
+  data-definition-for="concept-accept-ch-lifetime-cache">create a new entry</dfn> in the <span
```
"create" is not a good verb here, since usually creating doesn't have side effects. I'd suggest "add a new Accept-CH-Lifetime cache entry".

yoavweiss replied:
ack

domenic commented on source:
```diff
@@ -86177,8 +86292,59 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
+  <p>To <dfn data-x="accept-ch-lifetime-cache-create-entry"
+  data-definition-for="concept-accept-ch-lifetime-cache">create a new entry</dfn> in the <span
+  data-x="concept-accept-ch-lifetime-cache">Accept-CH-Lifetime cache</span>, given
+  <var>origin</var>, <var>max-age</var>, and <var>client-hints list</var>, do so as follows:
```
"do so as follows" followed by a DL is not so great. I'd suggest explicitly stating "add an entry with the following fields to the user agent's Accept-CH-Lifetime cache"

yoavweiss replied:
ack

domenic commented on source:
```diff
@@ -86177,8 +86292,59 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
+  </dl>
 
+  <p>To <dfn data-x="concept-retrieve-client-hints-list-accept-ch-lifetime-cache">retrieve
+  client-hints list</dfn>, given an <var>origin</var>:
```
retrieve _the_ client-hints list

No comma necessary

domenic commented on source:
```diff
@@ -86177,8 +86292,59 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
+  <p>To <dfn data-x="concept-retrieve-client-hints-list-accept-ch-lifetime-cache">retrieve
+  client-hints list</dfn>, given an <var>origin</var>:
+  <ol>
+  <li>Let <var>client-hints list</var> be an empty list.</li>
```
Missing indent

Missing `<p>` tags in each `<li>`

yoavweiss replied:
ack

domenic commented on source:
```diff
@@ -86177,8 +86292,59 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
+  client-hints list</dfn>, given an <var>origin</var>:
+  <ol>
+  <li>Let <var>client-hints list</var> be an empty list.</li>
+  <li>Fetch any entries from <span data-x="concept-accept-ch-lifetime-cache">Accept-CH-Lifetime
```
What does "Fetch" mean here?

I'd suggest "Let _originMatchingEntries_ be the entries in the ... whose origin is..."

yoavweiss replied:
makes sense

domenic commented on source:
```diff
@@ -86177,8 +86292,59 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
+  cache</span> which <span data-x="concept-accept-ch-lifetime-cache-origin"
+  data-definition-for="accept-ch-lifetime-cache">origin</span> is identical to <var>origin</var>.
+  </li>
+  <li>For each entry, for each token in its <span data-definitino-for="accept-ch-lifetime-cache">
```
"For each _entry_ of _originMatchingEntries_, for each _token_ of its client-hints list, append it to _client-hints-list_."

(By using https://infra.spec.whatwg.org/#set-append you won't need the extra check)

yoavweiss replied:
ack

domenic commented on source:
```diff
@@ -86177,8 +86292,59 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
+  <p>To <dfn data-x="concept-retrieve-client-hints-list-accept-ch-lifetime-cache">retrieve
+  client-hints list</dfn>, given an <var>origin</var>:
+  <ol>
+  <li>Let <var>client-hints list</var> be an empty list.</li>
```
Make client-hints list be an ordered set, both here and elsewhere, and you'll get the no-duplicates for free

yoavweiss replied:
makes sense

domenic commented on source:
```diff
@@ -86177,8 +86292,59 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
+  <p class="note">The Accept-CH lifetime cache can effectively act as an alternative cookie store,
+  since sites can use each of the hints as a bit set on the client, and that information will be
+  communicated to them on every request. As such, it is appropriate for user agents to evict that
+  cache whenever the user clears their cookies</p>
```
Should this be a normative "should" requirement or similar?

Also, add the fingerprint image, by using the `<!--INSERT FINGERPRINT-->`.

domenic commented on source:
```diff
@@ -14779,6 +14787,84 @@ people expect to have work and what is necessary.
+   <dd>
+
+    <p>This pragma sets the <span>environment settings object</span>'s
+    <span data-x="concept-settings-object-client-hints-list">client-hints list</span></p>
```
Missing period at end of sentence

domenic replied:
These descriptions are good, but it'd also be good to give slightly higher-level descriptions of the type that would be useful for someone just trying to get an idea of what these do. Compare to the other pragmas, and notice how they're mostly understandable to web developers.

domenic commented on source:
```diff
@@ -14779,6 +14787,84 @@ people expect to have work and what is necessary.
+     attribute, or if that attribute's value is the empty string, then return.</p></li>
+
+     <li><p>Let <var>callerIsSecureContext</var> be the result of executing <span>Is environment
+     settings object a secure context?</span> on the <code>Window</code> object's
```
"the Window object" is not well-defined here.

You probably want "the meta element's relevant settings object". And, save it since you'll use it later.

domenic commented on source:
```diff
@@ -78975,6 +79061,15 @@ interface <dfn>BarProp</dfn> {
+
+      <p>Return the result of running <span
+      data-x="concept-retrieve-client-hints-list-accept-ch-lifetime-cache">Retrieve client-hints
+      list</span> with <span data-x="concept-settings-object-origin">origin</span> as input.</p>
```
Probably "this environment settings object's origin" is clearer. And, no need for "as input".

domenic commented on source:
```diff
@@ -78975,6 +79061,15 @@ interface <dfn>BarProp</dfn> {
+     <dd>
+
+      <p>Return the result of running <span
+      data-x="concept-retrieve-client-hints-list-accept-ch-lifetime-cache">Retrieve client-hints
```
We don't capitalize algorithm names in HTML, generally.

domenic commented on source:
```diff
@@ -86486,6 +86652,13 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
+   <dt>A <dfn data-x="concept-settings-object-client-hints-list"
+   data-dfn-for="environment settings object">client-hints list</dfn></dt>
+
+   <dd><p>The <span data-x="concept-settings-object-client-hints-list">client-hints list</span> for
```
This link just links to itself, so this needs rephrasing. Maybe "A list of client hints used for fetches performed using..."?

yoavweiss replied:
ack

domenic commented on source:
```diff
@@ -82317,6 +82412,26 @@ interface <dfn>Location</dfn> { // but see also <a href="#the-location-interface
+     settings object a secure context?</span> on the <code>Window</code> object's
+     environment settings object.</p></li>
+
+     <li><p>If <var>response</var>'s
```
This model seems kind of broken.

What is a settings object's clients hints list? It's obtained by the algorithm "retrieve client-hints list", which creates a new list each time it's called. So, appending to it has no effect.

You seem to have mixed together two models which contradict each other:

- Client hints list is figured out from preexisting data ("retrieve client-hints list" figures out some subset of the Accept-CH lifetime cache that applies)
- Client hints list gets modified directly during Document creation

yoavweiss replied:
So the model I had in mind is:
* `Accept-CH` adds client hints to the current env settings object's client-hints list.
* After that, an `Accept-CH-lifetime` header stores entries in the ACHL cache, based on the current env settings object.
* When an env settings object is created, it's initial client-hints list is populated from the cache.

"retrieve client hints list" indeed creates a new list, but then populates it with of all cached hints from the cache (as there could be multiple entries which may correspond the current origin, with different hints on each.)

Does that make sense?

yoavweiss commented:
@domenic - thanks for reviewing! :) I addressed most of your comments, and left TODOs for the rest, which I intend to tackle a bit later (but before landing)

domenic commented:
(Continuing this discussion away from now-collapsed commit comments:)

> So the model I had in mind is:
> * `Accept-CH` adds client hints to the current env settings object's client-hints list.
> * After that, an `Accept-CH-lifetime` header stores entries in the ACHL cache, based on the current env settings object.
> * When an env settings object is created, it's initial client-hints list is populated from the cache.
> 
> "retrieve client hints list" indeed creates a new list, but then populates it with of all cached hints from the cache (as there could be multiple entries which may correspond the current origin, with different hints on each.)
> 
> Does that make sense?

It could make sense, but as written doesn't really work. Every time you say "some settings object's client-hints set", you run the algorithm specified in https://whatpr.org/html/3774/window-object.html#script-settings-for-window-objects , which re-creates a new list from scratch. So any additions you make to it (via Accept-CH) are only to that list, and are never seen again by any future callers of "some settings object's client-hints set".

If you want a single list that's created once, and then stored and returned later, you need to explicitly do that.

Relatedly, it seems a bit strange that you intend to create one client hints set per environment settings object (<-> Window), whereas all the other similar stuff like origin, HTTPS state, and referrer policy are associated with the Document. The main difference here is how you treat the transition from initial about:blank to a new Document within the same window. As written right now, if I add `<meta>` elements to the initial about:blank, then navigate to `https://example.com/`, the client hints that were set by those `<meta>` elements in `about:blank` will still apply to `https://example.com/`. That doesn't seem right to me?

I'd thus suggest a framework that more closely follows referrer policy or HTTPS state:

- Documents and WorkerGlobalScope instances get a client hints set, initially empty. (This is the only time we ever create client hints sets.)
- The environment settings object algorithms return that set. (The purpose of this is to provide a small indirection so that fetch only ever has to worry about clients = environment settings objects, not about Documents vs. WorkerGlobalScopes vs. ...)

Then I think this would work:

- When creating a document or workerglobalscope, "retrieve the client hints set" for the correct origin, and then loop over the result and append them to the document/WGS's client hints set.
- Accept-CH adds client hints to the Document/WGS's client hints set. (This now works, because running the "get client hints set" algorithm doesn't return a fresh set each time anymore, so mutating it has the desired effect.)

Thoughts?

yoavweiss commented:
@domenic makes perfect sense! Working on changing the model. Will be ready for re-review tomorrow.

yoavweiss commented:
@domenic technically, it's tomorrow for me :D I think I fixed all the comments. Can you PTAL?

I still need to update the corresponding Fetch PR, to rename its `client hints list` to `client hints set`

yoavweiss commented on source:
```diff
@@ -97614,6 +97877,14 @@ interface <dfn>SharedWorkerGlobalScope</dfn> : <span>WorkerGlobalScope</span> {
 
      </dd>
 
+     <dt>The <span data-x="concept-settings-object-client-hints-set">client-hints set</span></dt>
```
@domenic - I'm not sure this is what you meant by proxying the WGS's client-hints set through the settings object's client-hints set. Would that work?

domenic commented:
OK, I did a pass through for a final review that was mostly editorial. Please check my changes and let me know what you think.

I'd still prefer to hold off on this until @annevk has time to look at it from a higher conceptual/architectural. But, I think it's pretty solid at the level that I can understand.

yoavweiss commented on source:
```diff
@@ -86362,64 +86343,71 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
-  given <var>origin</var>, <var>max-age</var>, and <var>client-hints set</var>,
-  add a new entry with the following fields to the Accept-CH-Lifetime cache:
+  <p>To <dfn>add a new Accept-CH-Lifetime cache entry</dfn> to the <span
+  data-x="concept-accept-ch-lifetime-cache">Accept-CH-Lifetime cache</span>, given
```
Oops, changed the name here, but missed changing the link and the callers. Thanks for fixing!

yoavweiss commented:
@domenic -  Thanks for reviewing and fixing up. Your changes look good from my perspective. Waiting for @annevk to land the PRs makes sense.

annevk commented on source:
```diff
@@ -2791,6 +2791,7 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
      <li>A <dfn data-x-href="https://fetch.spec.whatwg.org/#network-scheme">network scheme</dfn></li>
      <li>A <dfn data-x-href="https://fetch.spec.whatwg.org/#fetch-scheme">fetch scheme</dfn></li>
      <li><dfn data-x-href="https://fetch.spec.whatwg.org/#concept-https-state-value">HTTPS state value</dfn></li>
+     <li><dfn data-x="concept-fetch-client-hints-set" data-x-href="https://fetch.spec.whatwg.org/#concept-client-hints-set">client-hints set</dfn></li>
```
It's unclear whether this should have a hyphen as discussed in the Fetch PR.

annevk replied:
I'd also suggest we drop data-x="" for this one, to ease reading.

yoavweiss replied:
Regarding the hyphen, I don't care much, tbh. Happy to remove it from both here and the Fetch spec if that makes things better.

Dropped the `data-x`.

yoavweiss replied:
oh, dropping the `data-x` causes the build to fail, so bringing it back

domenic replied:
If you drop the data-x then you would change all references to be `<span>client hints set</span>` instead of `<span data-x="concept-fetch-client-hints-set">client hints set</span>`. I think @annevk was suggesting that because it's simpler to read.

Sorry that Wattsi is a bit unintuitive at times. https://github.com/whatwg/wattsi/blob/master/Syntax.md if you haven't seen it.

annevk commented on source:
```diff
@@ -3945,6 +3946,26 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+
+    <ul class="brief">
+     <li><dfn data-x="concept-http-field-name" data-x-href="https://tools.ietf.org/html/rfc7230#section-3.2">field-name</dfn></li>
+     <li><dfn data-x="concept-http-delta-seconds" data-x-href="https://tools.ietf.org/html/rfc7234#section-1.2.1">delta-seconds</dfn></li>
```
We already reference HTTP for Content-Disposition and such. Shouldn't this be there? @domenic?

yoavweiss replied:
@domenic - thoughts if this should move somewhere else?

domenic replied:
Yeah, probably moving this to the "HTTP and related specifications" section is slightly nicer.

yoavweiss replied:
done

annevk commented on source:
```diff
@@ -14779,6 +14814,92 @@ people expect to have work and what is necessary.
+
+    <ol>
+     <li><p>If the <code>meta</code> element is not a child of a <code>head</code> element, then
+     return.</p></li>
```
Shouldn't this be more strict so it only works for head element that is the first element child of the root element? What do implementations do here? Do we test the difference?

yoavweiss replied:
Copied that piece from CSP. Maybe worth while to coalesce that logic into a single place. I don't think current tests make sure this (or CSP) only work for a head element that's the first child of root

annevk commented on source:
```diff
@@ -14779,6 +14814,92 @@ people expect to have work and what is necessary.
+     <code data-x="attr-meta-content">content</code> attribute's value.</p></li>
+
+     <li><p>Parse <var>acceptCHValue</var> according to the `<code>Accept-CH</code>` header parsing
+     rules, as a <span data-x="concept-http-field-name">field-name</span>. Append each parsed
```
Isn't it a field value?

yoavweiss replied:
It's defined as field-name at https://httpwg.org/http-extensions/client-hints.html#accept-ch
That makes sense as the values accepted are request header names

annevk replied:
So it's defined as a sequence of field-names, but that doesn't seem conveyed here? In particular the `#` is significant.

annevk commented on source:
```diff
@@ -14779,6 +14814,92 @@ people expect to have work and what is necessary.
+     <code data-x="attr-meta-content">content</code> attribute's value.</p></li>
+
+     <li><p>Parse <var>acceptCHValue</var> according to the `<code>Accept-CH</code>` header parsing
+     rules, as a <span data-x="concept-http-field-name">field-name</span>. Append each parsed
```
Should we link append?

yoavweiss replied:
linked to "set append"

annevk commented on source:
```diff
@@ -14779,6 +14814,92 @@ people expect to have work and what is necessary.
+     <code data-x="attr-meta-content">content</code> attribute's value.</p></li>
+
+     <li><p>Let <var>topLevelOrigin</var> be the <span>origin</span> of the <span>top-level
+     browsing context</span>'s <span>active document</span>.</p></li>
```
What is there's no browsing context?

What if there's an inbetween cross-origin `<iframe>`? I wonder if there's something this exposes that we're trying to prevent in #2480.

yoavweiss replied:
Have you got more details on the problem #2480 resolves?

annevk replied:
Preventing leaking the embedder to the embedded.

yoavweiss replied:
OK, the goal here is to prevent cross-origin iframes from settings values in the ACHL cache. I can't see how this would expose the embedder to the embedded, but I may be wrong. Is there some other/better way to prevent cross-origin iframes from setting ACHL entries which is safer?

annevk replied:
What does your implementation do?

yoavweiss replied:
I believe this is what the Chromium implementation [does](https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/loader/http_equiv.cc?q=http_equiv.cc&sq=package:chromium&dr&l=33).

/cc @tarunban

annevk replied:
Right, then yes, as ideally the standard matches the implementations.

tarunban replied:
Bailing out early sounds good.

yoavweiss replied:
Cool, bailing

annevk commented on source:
```diff
@@ -82317,6 +82447,41 @@ interface <dfn>Location</dfn> { // but see also <a href="#the-location-interface
+     data-x="concept-http-field-name">field-name</span>. Add each parsed client hints token to the
+     <var>Document</var>'s <span data-x="concept-document-client-hints-set">client-hints set</span>.
+     </p></li>
+
```
One enter too many.

yoavweiss replied:
ack

annevk commented on source:
```diff
@@ -82317,6 +82447,41 @@ interface <dfn>Location</dfn> { // but see also <a href="#the-location-interface
+     `<code>Accept-CH</code>` header parsing rules, as a <span
+     data-x="concept-http-field-name">field-name</span>. Add each parsed client hints token to the
+     <var>Document</var>'s <span data-x="concept-document-client-hints-set">client-hints set</span>.
+     </p></li>
```
You need to make it wrap as `set</span>.</p></li>` instead.

yoavweiss replied:
ok

annevk commented on source:
```diff
@@ -82317,6 +82447,41 @@ interface <dfn>Location</dfn> { // but see also <a href="#the-location-interface
+
+     <li><p>Let <var>topLevelOrigin</var> be the <span>origin</span> of the <span>top-level
+     browsing context</span>'s <span>environment settings object</span>'s <span>origin</span>.
+     </p></li>
```
Same wrapping issue.

yoavweiss replied:
cool

annevk commented on source:
```diff
@@ -86177,8 +86342,72 @@ interface <dfn>ApplicationCache</dfn> : <span>EventTarget</span> {
+  <p class="note">The Accept-CH lifetime cache can effectively act as an alternative cookie store,
+  since sites can use each of the hints as a bit set on the client, and that information will be
+  communicated to them on every request. As such, it is appropriate for user agents to evict that
+  cache whenever the user clears their cookies.
```
This should be a should requirement, not a note.

yoavweiss replied:
remove the note classifier
